### PR TITLE
[Merged by Bors] - Cleaner logic for gossip subscriptions for new forks

### PR DESCRIPTION
**Pull request by pawanjay176**
## Issue Addressed

Cleaner resolution for #4006 

## Proposed Changes

We are currently subscribing to core topics of new forks way before the actual fork since we had just a single `CORE_TOPICS` array. This PR separates the core topics for every fork and subscribes to only required topics based on the current fork.
Also adds logic for subscribing to the core topics of a new fork only 2 slots before the fork happens.

2 slots is to give enough time for the gossip meshes to form. 

Currently doesn't add logic to remove topics from older forks in new forks. For e.g. in the coupled 4844 world, we had to remove the `BeaconBlock` topic in favour of `BeaconBlocksAndBlobsSidecar` at the 4844 fork. It should be easy enough to add though. Not adding it because I'm assuming that  #4019 will get merged before this PR and we won't require any deletion logic. Happy to add it regardless though.

### CODE REVIEW

File: beacon_node/lighthouse_network/src/types/topics.rs [hunk 1]
```diff
@@ -23,21 +23,47 @@ pub const BLS_TO_EXECUTION_CHANGE_TOPIC: &str = "bls_to_execution_change";
...
+pub const CAPELLA_CORE_TOPICS: [GossipKind; 1] = [GossipKind::BlsToExecutionChange];
+
+pub const EIP4844_CORE_TOPICS: [GossipKind; 1] = [GossipKind::BeaconBlocksAndBlobsSidecar];
+
 pub const LIGHT_CLIENT_GOSSIP_TOPICS: [GossipKind; 2] = [
     GossipKind::LightClientFinalityUpdate,
     GossipKind::LightClientOptimisticUpdate,
 ];
 
+/// Returns the core topics associated with each fork
```
Comment (divagant-martian) [inline beacon_node/lighthouse_network/src/types/topics.rs]:
```suggestion
/// Returns the core topics associated with each fork that are new to the previous fork
```
Just a suggestion to make clear this is not the complete set of topics for the fork, but the new ones to add

File: beacon_node/lighthouse_network/src/types/topics.rs [hunk 2]
```diff
@@ -23,21 +23,47 @@ pub const BLS_TO_EXECUTION_CHANGE_TOPIC: &str = "bls_to_execution_change";
...
+/// Returns all the topics that we need to subscribe to for a given fork
+/// including topics from older forks and new topics for the current fork.
+pub fn core_topics_to_subscribe(current_fork: ForkName) -> Vec<GossipKind> {
+    let mut topics = fork_core_topics(&current_fork);
+    while let Some(previous_fork) = current_fork.previous_fork() {
+        let mut previous_fork_topics = fork_core_topics(&previous_fork);
+        topics.append(&mut previous_fork_topics);
+    }
+    topics
+}
```
Comment (divagant-martian) [inline beacon_node/lighthouse_network/src/types/topics.rs]: The idea sounds very elegant. I think you need to update the value of `current_fork` with that of `previous_fork` for this to go all the way back to genesis

**Review by divagant-martian — APPROVED** [review]

**Review by divagant-martian — CHANGES_REQUESTED** [review]
A minor suggestion for some docs and what I think is a bug in the logic but otherwise this sounds great to me, thanks a lot!

File: beacon_node/lighthouse_network/src/types/topics.rs (hunk 2 above)
Comment (pawanjay176) [reply beacon_node/lighthouse_network/src/types/topics.rs]: damn, i had fixed it after running a local testnet. Didn't commit. Thanks for catching it.
Added a small test for the function now.

**Review by divagant-martian — APPROVED** [review]
Looks solid to me, ran as expected in sepolia as well, Ty a lot @pawanjay176

**Comment by pawanjay176** [comment]
Tested this PR at the fork transition on sepolia (Thanks @paulhauner )
Mesh peers for `bls_to_execution_change` topic right before the fork on sepolia (everything except the white line was running this PR)

Screenshot from 2023-02-28 21-31-57

Fork happens at 09:34:48, we subscribed at 09:34:24. Mesh peer counts are 10, 10, 8, 5. So looks like 2 slots is sufficient time to build a healthy mesh.

**Comment by michaelsproul** [comment]
Needs an update now that the Capella-cleanup is merged

bors r-

**Pull request closed without merging**